File: web/src/lib/taskTiming.ts
import type { Task } from './types'

/**
 * Task timing measures, mirrored from the Go domain helpers
 * (internal/services/task/domain/task_timing.go). The two measures are kept
 * deliberately DISTINCT — the user was explicit that the span that matters most
 * is pickup/queue latency ("how long until an agent is aware of a task"), NOT
 * execution time:
 *
 *   - Pickup latency:   createdAt (queued) → startedAt (first claimed).
 *   - In-progress time: startedAt → completedAt (or now, while in flight).
 *
 * Both derive from timestamps that already exist on the task, so no new field /
 * migration is required. In-progress duration is wall-clock and does not
 * subtract blocked/in_review intervals; an "active-only" measure would need a
 * status-transition history, intentionally deferred for the MVP.
 *
 * This is the data foundation the metrics page and over-threshold notifications
 * build on. Functions take an explicit `now` (defaulting to Date.now()) so
 * callers can render deterministically and tests stay stable.
 */

/** Parse an ISO timestamp to epoch ms, or null when absent/unparseable. */
function epochMs(iso?: string): number | null {
  if (!iso) return null
  const t = new Date(iso).getTime()
  return Number.isNaN(t) ? null : t
}

/** Terminal statuses — work is done, so they can never be "over threshold". */
const TERMINAL_STATUSES = new Set(['succeeded', 'failed', 'canceled'])

/**
 * Pickup latency in ms: startedAt - createdAt. null when either timestamp is
 * missing (never claimed, or unknown creation time). Clock skew is clamped to 0.
 */
export function pickupLatencyMs(task: Task): number | null {
  const created = epochMs(task.createdAt)
  const started = epochMs(task.startedAt)
  if (created === null || started === null) return null
  return Math.max(0, started - created)
}

/**
 * In-progress duration in ms: (completedAt ?? now) - startedAt. null when the
 * task was never started. Clock skew is clamped to 0.
 */
export function inProgressDurationMs(task: Task, now: number = Date.now()): number | null {
  const started = epochMs(task.startedAt)
  if (started === null) return null
  const completed = epochMs(task.completedAt)
  const end = completed ?? now
  return Math.max(0, end - started)
}

/**
 * Whether a still-in-flight task has been in progress past `thresholdMs`.
 * Terminal tasks never breach, an unstarted task never breaches, and a
 * non-positive threshold disables the check.
 */
export function isOverThreshold(
  task: Task,
  thresholdMs: number,
  now: number = Date.now(),
): boolean {
  if (thresholdMs <= 0 || TERMINAL_STATUSES.has(task.status)) return false
  const d = inProgressDurationMs(task, now)
  return d !== null && d > thresholdMs
}

/**
 * Coarse, human-readable duration for timing figures ("<1m", "5m", "2h 15m",
 * "3d 4h"). Distinct from format.ts `formatDuration`, which keeps sub-minute
 * precision for short execution spans — pickup latency and in-progress duration
 * read better at minute granularity. Returns null when given null.
 */
export function formatCoarseDuration(milliseconds: number | null): string | null {
  if (milliseconds === null) return null
  const mins = Math.floor(milliseconds / 60_000)
  if (mins < 1) return '<1m'
  if (mins < 60) return `${mins}m`
  const hrs = Math.floor(mins / 60)
  const remMins = mins % 60
  if (hrs < 24) return remMins > 0 ? `${hrs}h ${remMins}m` : `${hrs}h`
  const days = Math.floor(hrs / 24)
  const remHrs = hrs % 24
  return remHrs > 0 ? `${days}d ${remHrs}h` : `${days}d`
}

File: internal/services/task/domain/task_timing.go
package domain

import "time"

// Task timing measures, kept deliberately DISTINCT (per the agent-management
// mission KR). The user was explicit that the span that matters most is "how
// long it takes for an agent to be aware of a task" — pickup/queue latency —
// NOT execution time. We expose both, separately:
//
//   - Pickup latency: CreatedAt (queued) → StartedAt (first claimed). How long
//     work sat unpicked. This is the "time to awareness" figure.
//   - In-progress duration: StartedAt → CompletedAt (or now, while in flight).
//     How long work has been open since it began.
//
// Both derive from timestamps that already exist on the Task — verified against
// the lifecycle before adding any field, which is why this needs no migration:
//   - CreatedAt  is set at creation (the queued moment).
//   - StartedAt  is stamped once in Claim() on the first pending→active move and
//     never overwritten (Release keeps it; a re-claim is guarded by nil-check),
//     so it is a stable "work first began" marker.
//   - CompletedAt is stamped on the terminal transitions (approve/fail/cancel).
//
// In-progress duration is WALL-CLOCK and does not subtract blocked/in_review
// intervals. A precise "active-only" measure would require a status-transition
// history, which is intentionally deferred — the MVP signal the user asked for
// ("a task worked past a threshold of time, I want to know") is well served by
// elapsed-since-start, and the cost of a transition log isn't justified yet.

// PickupLatency returns StartedAt - CreatedAt: how long the task waited from
// being queued until it was first claimed. Returns nil when either timestamp is
// absent (never claimed, or unknown creation time). Clock skew that would yield
// a negative span is clamped to zero.
func (t Task) PickupLatency() *time.Duration {
	if t.CreatedAt == nil || t.StartedAt == nil {
		return nil
	}
	d := t.StartedAt.Sub(*t.CreatedAt)
	if d < 0 {
		d = 0
	}
	return &d
}

// InProgressDuration returns (CompletedAt ?? now) - StartedAt: how long the task
// has been in flight since work began. For a finished task the closed span uses
// CompletedAt; for a task still in progress the open end is `now`. Returns nil
// when the task was never started. A clock is passed in so the value is testable
// and deterministic. Negative spans (skew) are clamped to zero.
func (t Task) InProgressDuration(now time.Time) *time.Duration {
	if t.StartedAt == nil {
		return nil
	}
	end := now
	if t.CompletedAt != nil {
		end = *t.CompletedAt
	}
	d := end.Sub(*t.StartedAt)
	if d < 0 {
		d = 0
	}
	return &d
}

// IsOverThreshold reports whether a still-in-flight task has been in progress
// longer than `threshold` — the "agent over its time threshold on a task"
// signal. Terminal tasks never breach (the work is done), a task that never
// started never breaches, and a non-positive threshold disables the check.
func (t Task) IsOverThreshold(threshold time.Duration, now time.Time) bool {
	if threshold <= 0 || t.IsTerminal() {
		return false
	}
	d := t.InProgressDuration(now)
	if d == nil {
		return false
	}
	return *d > threshold
}
